ergm.py

```diff
@@ -187,6 +201,127 @@ def sample_network(self, sampling_method="NaiveMetropolisHastings", seed_network
+        else:
+            upper_triangle_indices = np.triu_indices(num_nodes, k=1)
+            cur_adj_mat[upper_triangle_indices] = mat_entries_arr
+            lower_triangle_indices_aligned = (upper_triangle_indices[1], upper_triangle_indices[0])
+            cur_adj_mat[lower_triangle_indices_aligned] = mat_entries_arr
+        return cur_adj_mat
+
+    @staticmethod
+    def construct_int_from_adj_mat(adj_mat: np.ndarray, is_directed: bool) -> int:
+        if len(adj_mat.shape) != 2 or adj_mat.shape[0] != adj_mat.shape[1]:
+            raise ValueError(f"The dimensions of the given adjacency matrix {adj_mat.shape} are not valid for an "
+                             f"adjacency matrix (should be a 2D squared matrix)")
+        num_nodes = adj_mat.shape[0]
+        adj_mat_no_diag = adj_mat[~np.eye(num_nodes, dtype=bool)]
+        if not is_directed:
+            adj_mat_no_diag = adj_mat_no_diag[:adj_mat_no_diag.size // 2]
+        return round((adj_mat_no_diag * 2 ** np.arange(adj_mat_no_diag.size - 1, -1, -1).astype(np.ulonglong)).sum())
+
+    def __init__(self,
+                 n_nodes,
+                 network_statistics: NetworkStatistics,
+                 is_directed=False,
+                 initial_thetas=None,
+                 initial_normalization_factor=None,
+                 seed_MCMC_proba=0.25):
+        super().__init__(n_nodes,
+                         network_statistics,
+                         is_directed,
+                         initial_thetas,
+                         initial_normalization_factor,
+                         seed_MCMC_proba)
+        self._all_weights = self._calc_all_weights()
+        self._normalization_factor = self._all_weights.sum()
+
+    # TODO: think of an alternative name? confusing with the static method
+    def _get_adj_mat_from_int(self, int_code: int):
+        return BruteForceERGM.construct_adj_mat_from_int(int_code, self._n_nodes, self._is_directed)
+
+    def _calc_all_weights(self):
+        if self._n_nodes > BruteForceERGM.MAX_NODES_BRUTE_FORCE:
+            raise ValueError(
+                f"The number of nodes {self._n_nodes} is larger than the maximum allowed for brute force "
+                f"calculations {BruteForceERGM.MAX_NODES_BRUTE_FORCE}")
+        num_pos_connects = self._n_nodes * (self._n_nodes - 1)
+        if not self._is_directed:
+            num_pos_connects //= 2
+        space_size = 2 ** num_pos_connects
+        all_weights = np.zeros(space_size)
+        for i in range(space_size):
+            cur_adj_mat = self._get_adj_mat_from_int(i)
+            all_weights[i] = super().calculate_weight(cur_adj_mat)
+        return all_weights
+
+    def calculate_weight(self, W: np.ndarray):
+        adj_mat_idx = BruteForceERGM.construct_int_from_adj_mat(W, self._is_directed)
+        return self._all_weights[adj_mat_idx]
+
+    def sample_network(self, sampling_method="Exact", seed_network=None, steps=0):
```

Comment:
why do you need a sample_network function in the BruteforceERGM class? doesn't seem to be used anywhere

Reply:
I didn't pay attention to the fact it is not in use...
Why shouldn't we implement it? Its logic is different than that of the parent ERGM class. If we have a class of exact calculations for small nets, doesn't it make sense to support the API of the parent class in full?